### DISCUSSION TITLE
Crop an image based on one channel

### DISCUSSION THREAD

**Discussion by cskilbeck** [ImageMagick/ImageMagick#4243] [Help] [answered]
Hi, I have an image (see below) with a border in the red channel and the stuff I'd like to keep in the green channel.

I'd like to:
- crop the image to the extent of the red pixels (ignoring any which aren't 100% red)
- crop N more pixels (4, say) from each side
- delete the red channel
- copy the green channel to the red and blue channels (making it white)
- save the result

I've been reading the docs but am stumped, can anyone with more experience of this program help me?

I'm using windows, version: ImageMagick 7.1.0-7 Q16-HDRI x64 2021-09-12

Thanks,
Charlie.

![small_test1](https://user-images.githubusercontent.com/4994859/133824793-4758f214-9570-441f-acb3-985c4e061ab5.png)

**Comment by snibgo** [comment]
Windows BAT syntax:
```
magick ^
  jlcjlc.png ^
  -strip ^
  -trim ^
  -shave 4x4 +repage ^
  -channel G -separate +channel ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/133829295-72a74ca9-e94c-4ab7-ba29-fc085c685550.png)

"-strip" removes the embedded sRGB profile. This is arguable; it may be better to leave it.

Note that this has trimmed off green elements at the bottom.

**Comment by snibgo** [comment, marked as answer]
Sorry, the above isn't exactly what you asked for. Your first step...

> crop the image to the extent of the red pixels (ignoring any which aren't 100% red)

... is a bit tricky. My solution is to make a clone, isolate the red channel, and find what "-trim" would do. Discard the clone, and apply that crop to the original image.

```
magick ^
  jlcjlc.png ^
  -strip ^
  ( +clone ^
    -channel R -separate +channel ^
    -set option:MYCROP "%%@" ^
    +delete ^
  ) ^
  -crop %%[MYCROP] +repage ^
  -shave 4x4 +repage ^
  -channel G -separate +channel ^
  x.png
```

**Reply by cskilbeck** [reply]
ooh, you're a genius, it's almost perfect, I just need to crop everything where the red channel is not 100% (255). I think as it is it crops to any pixels that have anything non-zero in the red channel - does that make sense?

**Reply by fmw42** [reply]
Threshold the red channel to isolate the 100% red pixels. Then crop.

**Reply by cskilbeck** [reply]
Yep, this seems to work
```
c:\bin\ImageMagick\magick.exe ^
 %1 ^
 -strip ^
 ( +clone ^
 -channel R -separate +channel ^
 -color-threshold "red-red" ^
 -set option:MYCROP "%%@" ^
 +delete ^
 ) ^
 -crop %%[MYCROP] +repage ^
 -shave 8x8 +repage ^
 -channel G -separate +channel ^
 %1_cropped.png
```

**Reply by cskilbeck** [reply]
Thanks, I've been tearing my hair out - you're a life saver

**Comment by snibgo** [comment]
After "-channel R -separate +channel", the image is grayscale. Every pixel is white or black or a shade of gray. So "-color-threshold red-red" is misleading, because no pixel is red. It happens to work, but I'm not sure why.

But it should work and does work if we remove the separation of the red channel. Then "-color-threshold" works on an image that does have red pixels. It changes those to white, and and the rest to black.
```
magick ^
  jlcjlc.png ^
  -strip ^
  ( +clone ^
    -color-threshold "red-red" ^
    -set option:MYCROP "%%@" ^
    +delete ^
  ) ^
  -crop %%[MYCROP] +repage ^
  -shave 4x4 +repage ^
  -channel G -separate +channel ^
  x.png
```
